geoclustering/triclustering_numpy.py

```diff
@@ -0,0 +1,104 @@
+import numpy as np
+
+
+def _distance(Z, X, Y, epsilon):
+    """ Distance function """
+    Y = Y + epsilon
+    d = np.dot(X, Y) - np.dot(Z, np.log(Y))
+    return d
+
+
+def _initialize_clusters(n_el, n_clusters):
+    """ Initialize cluster occupation matrix """
+    cluster_idx = np.mod(np.arange(n_el), n_clusters)
+    cluster_idx = np.random.permutation(cluster_idx)
+    eye = np.eye(n_clusters, dtype=np.int32)
+    return eye[cluster_idx]
+
+
+def triclustering(Z, k, l, z, errobj, niters, epsilon):
+    """
+    Run the co-clustering, Numpy-based implementation
+
+    :param Z: (d x m x n) data matrix
+    :param k: num row clusters
+    :param l: num col clusters
+    :param z: num band clusters
+    :param errobj: precision of obj fun for convergence
+    :param niters: max iterations
+    :param epsilon: precision of matrix elements
+    :return: has converged, number of iterations performed. final row, column
+    and band clustering, error value
+    """
+    [d, m, n] = Z.shape
+
+    # get data matrix size and fixed averages(upload rows)
+    Y = np.concatenate(Z, axis=1)
+    Gavg = Y.mean()
+
+    # get data matrix size and fixed averages(upload columns)
+    Y1 = np.concatenate(Z, axis=0)
+    Gavg1 = Y1.mean()
+
+    # get data matrix size and fixed averages(upload 3D)
+    Y2 = Z.reshape(d, m*n)
+    Gavg2 = Y2.mean()
+
+    # Randomly initialize row and column clustering
+    R = _initialize_clusters(m, k)
+    C1 = _initialize_clusters(n*d, z)
+    C = _initialize_clusters(n, l)
+    Z = _initialize_clusters(d, z)
```

Comment:
Consider not overwriting the input matrix Z

Reply:
Thanks for catching this! Bad namings on my side!